funflow/src/Control/FunFlow/ContentStore.hs

```diff
@@ -395,17 +406,36 @@ waitUntilComplete store hash = lookupOrWait store hash >>= \case
 
 -- | Atomically query the state under the given key and mark pending if missing.
 -- Return an 'Control.Concurrent.Async' to await updates, if already pending.
-constructOrWait
+constructOrAsync
   :: ContentStore
   -> ContentHash
   -> IO (Status (Path Abs Dir) (Async Update) Item)
-constructOrWait store hash = withStoreLock store $
+constructOrAsync store hash = withStoreLock store $
   internalQuery store hash >>= \case
     Complete item -> return $ Complete item
     Missing () -> withWritableStore store $
       Missing <$> createBuildDir store hash
     Pending _ -> Pending <$> internalWatchPending store hash
 
+-- | Atomically query the state under the given key and mark pending if missing.
+-- Wait for the item to be completed, if already pending.
+-- Throws a 'FailedToConstruct' error if construction fails.
+constructOrWait
+  :: ContentStore
+  -> ContentHash
+  -> IO (Status (Path Abs Dir) Void Item)
+constructOrWait store hash = constructOrAsync store hash >>= \case
+  Pending a -> wait a >>= \case
+    Completed item -> return $ Complete item
+    -- XXX: Consider extending 'Status' with a 'Failed' constructor.
+    --   If the store contains metadata as well, it could keep track of the
+    --   number of failed attempts and further details about the failure.
+    --   If an external task is responsible for the failure, the client could
+    --   choose to resubmit a certain number of times.
+    Failed -> throwIO $ FailedToConstruct hash
+  Complete item -> return $ Complete item
+  Missing dir -> return $ Missing dir
```

Comment:
The semantics of this seem slightly confusing - e.g. the directory is no longer missing, since we have created it at https://github.com/tweag/funflow/pull/42/files#diff-06fe0eceda2b774b025d64b91b0c4f0cR417. Possibly expand the comments to detail what the various return cases mean?

Reply:
I've added some explanation in the haddocks.